src/libxtp/davidsonsolver.cc

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright 2009-2018 The VOTCA Development Team (http://www.votca.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <stdexcept>
+#include <iostream>
+
+#include <Eigen/Dense>
+#include <Eigen/Core>
+#include <Eigen/QR>
+#include <Eigen/Eigenvalues>
+#include <Eigen/IterativeLinearSolvers>
+#include <unsupported/Eigen/IterativeSolvers>
+#include <chrono>
+
+#include <votca/xtp/matrixfreeoperator.h>
+#include <votca/xtp/bse_matrix_free.h>
+#include <votca/xtp/davidsonsolver.h>
+
+using boost::format;
+using std::flush;
+
+namespace votca { namespace xtp {
+
+using namespace std;
+
+DavidsonSolver::DavidsonSolver(ctp::Logger &log) : _log(log), iter_max(1000), tol(1E-6), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max) : iter_max(iter_max) , tol(1E-6), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max, real_gwbse tol) : iter_max(iter_max), tol(tol), max_search_space(100) { }
+// DavidsonSolver::DavidsonSolver(int iter_max, real_gwbse tol, int max_search_space) : iter_max(iter_max), tol(tol), max_search_space(max_search_space) { }
+
+void DavidsonSolver::set_iter_max(int N) { this->iter_max = N; }
+void DavidsonSolver::set_tolerance(real_gwbse eps) { this->tol = eps; }
+void DavidsonSolver::set_max_search_space(int N) { this->max_search_space = N;}
+void DavidsonSolver::set_jacobi_correction() { this->jacobi_correction = true; }
+void DavidsonSolver::set_jacobi_linsolve(int method) {this->jacobi_linsolve = method;}
+
+VectorXfd DavidsonSolver::eigenvalues() {return this->_eigenvalues;}
+MatrixXfd DavidsonSolver::eigenvectors() {return this->_eigenvectors;}
```

Comment:
So either write davidson in general for the templated versions but please do not write it for MatrixXfd but stay in double